ARROW-8562: [C++] IO: Parameterize I/O Coalescing using S3 metrics

mayuropensource opened the pull request:
_(Recreating the PR from a clean repo, sorry about earlier PR which was not cleanly merged)._

**JIRA:** https://issues.apache.org/jira/browse/ARROW-8562

This change is not actually used until #6744 (@lidavidm) is pushed, however, it doesn't need to wait for the other pull request to be merged.

**Description:**
The adaptive I/O coalescing algorithm uses two parameters:

    max_io_gap or hole_size_limit: Max I/O gap/hole size in bytes
    ideal_request_size or range_size_limit: Ideal I/O Request size in bytes

These parameters can be derived from S3 metrics as described below:

In an S3 compatible storage, there are two main metrics:

    Seek-time or Time-To-First-Byte (TTFB) in seconds: call setup latency of a new S3 request

    Transfer Bandwidth (BW) for data in bytes/sec

    Computing max_io_gap or hole_size_limit:

max_io_gap = TTFB * BW

This is also called Bandwidth-Delay-Product (BDP).

Two byte ranges that have a gap can still be mapped to the same read if the gap is less than the bandwidth-delay product [TTFB * TransferBandwidth], i.e. if the Time-To-First-Byte (or call setup latency of a new S3 request) is expected to be greater than just reading and discarding the extra bytes on an existing HTTP request.

    Computing ideal_request_size or range_size_limit:

We want to have high bandwidth utilization per S3 connections, i.e. transfer large amounts of data to amortize the seek overhead.
But, we also want to leverage parallelism by slicing very large IO chunks. We define two more config parameters with suggested default values to control the slice size and seek to balance the two effects with the goal of maximizing net data load performance.

BW_util (ideal bandwidth utilization):
This means what fraction of per connection bandwidth should be utilized to maximize net data load.
A good default value is 90% or 0.9.

MAX_IDEAL_REQUEST_SIZE:
This means what is the maximum single request size (in bytes) to maximize net data load.
A good default value is 64 MiB.

The amount of data that needs to be transferred in a single S3 get_object request to achieve effective bandwidth eff_BW = BW_util * BW is as follows:
eff_BW = ideal_request_size / (TTFB + ideal_request_size / BW)

Substituting TTFB = max_io_gap / BW and eff_BW = BW_util * BW, we get the following result:
ideal_request_size = max_io_gap * BW_util / (1 - BW_util)

Applying the MAX_IDEAL_REQUEST_SIZE, we get the following:
ideal_request_size = min(MAX_IDEAL_REQUEST_SIZE, max_io_gap * BW_util / (1 - BW_util))

lidavidm commented on cpp/src/arrow/io/caching.h:
```diff
@@ -27,6 +27,44 @@
+  ///   to maximize the net data load.
+  ///   The value is a positive integer.
+  /// \return A new instance of CacheOptions.
+  static CacheOptions MakeFromS3Metrics(
```
This might be better named 'NetworkMetrics' or something along those lines; it's not specific to S3 and would be likely be useful for the in-progress Google Cloud Storage filesystem, and perhaps others.

mayuropensource replied:
good point, I will make the change.

fsaintjacques commented:
Could you accompany a script/utility to compute both metrics? Paired with toxiproxy, we could replicate S3 regions behavior with localhost.

mayuropensource commented:
@fsaintjacques, I can try to put together a python script using boto to determine the S3 metrics. Will that work for you?

fsaintjacques commented:
I'd say just plain HTTP, as @lidavidm pointed in his comment, this is a network attribute.

mayuropensource commented:
// SOME_S3_DATA_URI should point to a file (over http) that is ~500 MiB.
// TTFB_sec is the time-to-first-byte in seconds as measured by curl
// Bandwidth_bytes_per_sec is download bandwidth (in bytes per second) as measured by curl

`S3_DATA_URI="SOME_S3_DATA_URI"; `

`curl --negotiate -u: -o /dev/null -w "connect_time_sec=%{time_connect} total_time_sec=%{time_total} data_size_bytes=%{size_download} TTFB_sec=%{time_starttransfer} Bandwidth_bytes_per_sec=%{speed_download}\n" $S3_DATA_URI`

mayuropensource commented:
A better calculation for bandwidth (by removing TTFB from total time) is done using following script:

curl --negotiate -u: -o /dev/null -w "total_time_sec=%{time_total} data_size_bytes=%{size_download} TTFB_sec=%{time_starttransfer}\n" $S3_DATA_URI | awk -F' ' '{ split($1, tt, "="); split($2, ds, "="); split($3, ttfb, "="); bw=ds[2] / (1024 * 1024. * (tt[2] - ttfb[2])); print("TTFB_millis =", ttfb[2] * 1000.); print("Bandwidth_MiB_per_sec =", bw); }'

mayuropensource commented:
thank you @wesm
